### DISCUSSION TITLE
Invalid parameter: messages with role 'tool' must be a response to a preceeding message with 'tool_calls' .

### DISCUSSION THREAD

**Discussion by msajjadariqt** [microsoft/semantic-kernel#8419] [Q&A] [answered]
Can someone please tell me where can be the issue which is causing this issue?

`
public async Task<string> ProcessUserQuery(UserQueryRequest queryRequest)
{
    _logger.LogInformation("Received user query: {Query}, in method: {MethodName}", queryRequest.Query, nameof(ProcessUserQuery));

    try
    {
        return await ExecuteWithRetryAsync(async () =>
        {
            // Use the reusable method
            string resultContent = await ExecuteChatCompletionAsync(
                chatHistory: _chatHistory,
                query: queryRequest.Query,
                authorRole: AuthorRole.User,
                settings: new OpenAIPromptExecutionSettings { ToolCallBehavior = ToolCallBehavior.AutoInvokeKernelFunctions }
            );

            // Additional logic if needed
            if (ContainsRelevantKeywords(resultContent))
            {
                string additionalInfo = await PerformAutoCrawling(resultContent);
                resultContent = $"{resultContent}\n\nAdditional information gathered:\n{additionalInfo}";
            }

            return resultContent;

        }, maxRetryAttempts, delayBetweenRetries);
    }
    catch (Exception ex)
    {
        _logger.LogError(ex, "Error processing user query");
        return "An error occurred while processing your request. Please try again later.";
    }
}

private static void ManageChatHistory(ChatHistory _chatHistory, string query, AuthorRole authorRole)
{
    if (_chatHistory.Count > maxChatHistoryLength)
    {
        _chatHistory.RemoveRange(0, _chatHistory.Count - maxChatHistoryLength); // Remove all excess records at once
    }

    if (authorRole == AuthorRole.User)
    {
        _chatHistory.AddUserMessage(query);
    } 
    _chatHistory.AddAssistantMessage(query);
}

private async Task<string> ExecuteWithRetryAsync(Func<Task<string>> action, int maxRetryAttempts, TimeSpan delayBetweenRetries)
{
    int currentAttempt = 0;
    while (currentAttempt < maxRetryAttempts)
    {
        try
        {
            currentAttempt++;
            // Attempt to execute the action
            var result = await action();

            // Check if the result is valid
            if (!string.IsNullOrEmpty(result))
            {
                return result; // Return result if successful
            }
            else
            {
                _logger.LogWarning("Attempt {AttemptNumber} failed due to no valid result from tool. Retrying...", currentAttempt);
            }
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Attempt {AttemptNumber} failed. Retrying...", currentAttempt);

            if (currentAttempt >= maxRetryAttempts)
            {
                _logger.LogError(ex, "Max retry attempts reached.");
                return "An error occurred while processing your request after several attempts. Please try again later." + ex.Message.ToString();
            }

            await Task.Delay(delayBetweenRetries); // Optional delay before retrying
        }
    }

    return "An unexpected error occurred."; // Fallback in case something goes wrong in the logic.
}
private async Task<string> PerformAutoCrawling(string query)
{
    // Implement logic to gather more information from the internet or kernel functions
    query = "Gather additional from the internet, chat history or data sources, without specifying internet sources: ";

    // Use the reusable method
    return await ExecuteChatCompletionAsync(
        chatHistory: _chatHistory,
        query: query,
        authorRole: AuthorRole.User,
        settings: new OpenAIPromptExecutionSettings { ToolCallBehavior = ToolCallBehavior.AutoInvokeKernelFunctions }
    );
}

private static bool ContainsRelevantKeywords(string content)
{
    var words = content.Split(' ', StringSplitOptions.RemoveEmptyEntries);
    foreach (var word in words)
    {
        if (LogKeywordConfig.Keywords.Contains(word))
        {
            return true;
        }
    }
    return false;
}

private async Task<string> ExecuteChatCompletionAsync(ChatHistory chatHistory, string query, AuthorRole authorRole, OpenAIPromptExecutionSettings settings)
{
    ManageChatHistory(chatHistory, query, authorRole);

    var result = await _chatCompletionService.GetChatMessageContentAsync(
        chatHistory: chatHistory,
        executionSettings: settings,
        kernel: _kernel);

    if (result != null && !string.IsNullOrEmpty(result.Content?.ToString()))
    {
        ManageChatHistory(chatHistory, result.Content.ToString(), AuthorRole.Assistant);
        return result.Content.ToString();
    }

    return "No data available for the query.";
}`

**Comment by moonbox3** [comment, marked as answer]
Hi @msajjadariqt, the error in your subject "messages with role `tool` must be a response to a preceding message with `tool_calls`" occurs because the chat history seems to be misconfigured. When dealing with tool calls, and the model, the chat history needs to have the messages in the proper order (with the proper roles) otherwise it throws this error.

For example, with function calling enabled: if a user asks the model a question: `What is 10101+1001?` the model returns a finish_reason=`tool_calls`, and then it expects the next message to contain the tool_calls response(s) with the role='tool'. You can see this represented in the screenshot (you'll want to click on it so you can zoom in).

![image](https://github.com/user-attachments/assets/ab28d64b-7005-4942-a9b6-42dce95a49b0)

Can you please double check what you have included as part of your chat history? You may be missing adding the response from the model that contains the tool_calls that are processed in SK.

**Reply by msajjadariqt** [reply]
**Can you please double check what you have included as part of your chat history? You may be missing adding the response from the model that contains the tool_calls that are processed in SK.**

We don't manually add the tool_calls and we are only adding the _chatHistory.AddUserMessage(query);

I have multiple kernal functions which are being called for the prompt. Does failing of any function results in the tool_calls error?
